tools/fp20compiler/rc1.0_combiners.cpp

```diff
@@ -21,37 +21,47 @@ void CombinersStruct::Validate()
 
 void CombinersStruct::Invoke()
 {
-    // for (int i = 0; i < numConsts; i++)
-    //     glCombinerParameterfvNV(cc[i].reg.bits.name, &(cc[i].v[0]));
     assert(numConsts <= 2);
-
-    printf("pb_push1(p, NV097_SET_COMBINER_CONTROL,");
-    printf("\n    MASK(NV097_SET_COMBINER_CONTROL_FACTOR0, %s)",
-        numConsts >= 1 ? "NV097_SET_COMBINER_CONTROL_FACTOR1_SAME_FACTOR_ALL"
-            : "NV097_SET_COMBINER_CONTROL_FACTOR0_EACH_STAGE");
-    printf("\n    | MASK(NV097_SET_COMBINER_CONTROL_FACTOR1, %s)",
-        numConsts >= 2 ? "NV097_SET_COMBINER_CONTROL_FACTOR1_SAME_FACTOR_ALL"
-            : "NV097_SET_COMBINER_CONTROL_FACTOR1_EACH_STAGE");
-    printf("\n    | MASK(NV097_SET_COMBINER_CONTROL_ITERATION_COUNT, %d)", generals.num);
-    printf(");\n");
-    printf("p += 2;\n");
-
     for (int i = 0; i < numConsts; i++) {
-        const char* cmd = NULL;
-        switch (cc[i].reg.bits.name) {
+    //     glCombinerParameterfvNV(cc[i].reg.bits.name, &(cc[i].v[0]));
+        const char* general_cmd = NULL;
+        const char* final_cmd = NULL;
+        switch(cc[i].reg.bits.name) {
         case REG_CONSTANT_COLOR0:
-            cmd = "NV097_SET_SPECULAR_FOG_FACTOR + 0";
+            general_cmd = "NV097_SET_COMBINER_FACTOR0";
+            final_cmd = "NV097_SET_SPECULAR_FOG_FACTOR + 0";
             break;
         case REG_CONSTANT_COLOR1:
-            cmd = "NV_097_SET_SPECULAR_FOG_FACTOR + 4";
+            general_cmd = "NV097_SET_COMBINER_FACTOR1";
+            final_cmd = "NV097_SET_SPECULAR_FOG_FACTOR + 4";
             break;
         default:
             assert(false);
             break;
         }
-        printf("pb_push4f(p, %s, %f, %f, %f, %f);\n", cmd,
-            cc[i].v[0], cc[i].v[1], cc[i].v[2], cc[i].v[3]);
-        printf("p += 5;\n");
+
+        assert(cc[i].v[0] >= 0.0f && cc[i].v[0] <= 1.0f);
+        assert(cc[i].v[1] >= 0.0f && cc[i].v[1] <= 1.0f);
+        assert(cc[i].v[2] >= 0.0f && cc[i].v[2] <= 1.0f);
+        assert(cc[i].v[3] >= 0.0f && cc[i].v[3] <= 1.0f);
+
+        if (generals.localConsts == 0) {

```

Comment:
Can also be skipped if no general-combiners exist

Reply:
This is impossible (= hard) to check for because of `// Setup a "fake" general combiner 0` forces `generals.num` to 1 (when it was `0`) - this implies the hardware needs at least one iteration.

Reply:
Yes, hardware needs at least 1 stage.